cis194/week3/zehua/src/Exercises.hs

```diff
@@ -0,0 +1,80 @@
+module Exercises where
+
+import           Control.Monad (ap)
+import           Data.Char (isSpace)
+import           Data.List (sort, transpose)
+-- import           Data.Tuple.Utils -- from stackage
+
+-- ex1
+-- 117 chars
+
+-- helper function adapted from https://stackoverflow.com/a/2028218
+-- doing `drop n xs` instead of `drop (n-1) xs` to save 4 chars :)
+--   which requires passing in 0-based `zip [0..]` instead of `zip [1..]`
+-- keeping the name `every` instead of a short one-character name, e.g., `e`
+every :: (Int, [a]) -> [a]
+every (n, xs) = case drop n xs of
+              (y:ys) -> y : every (n, ys)
+              []     -> []
+
+skips :: [a] -> [[a]]
+-- infinite list first then cut off
+-- skips = takeWhile (not . null) . map every . zip [0..] . repeat
+-- dropWhileEnd does not seem to work on infinite list :|
+-- skips = dropWhileEnd null . map every . zip [0..] . repeat
+
+-- <*> S combinator on reader monad to implement s x y z = (x z) (y z)
+--   https://kseo.github.io/posts/2016-12-24-reader-monad-and-ski-combinators.html
+-- skips = map every . zip [0..] . (flip replicate <*> length)
+-- skips = take . length <*> map every . zip [0..] . repeat
+
+-- seems to be the most readable
+skips l = map every . zip [0..] $ replicate (length l) l
+
+
+-- ex2
+-- 83 chars
+
+localMaxima :: [Integer] -> [Integer]
+-- using drop avoids having to check the edge cases of length <= 2
+--localMaxima []    = []
+--localMaxima [_]   = []
+--localMaxima [_,_] = []
+localMaxima =
+    -- snd3 from Data.Tuple.Utils is not from haskell-platform :|
+    map (\(_, x, _) -> x)
+    . filter (\(x, y, z) -> x < y && y > z)
+    -- $ zip3 l (drop 1 l) (drop 2 $ l) l
+    . (zip3 <$> id <*> drop 1 <*> drop 2)
```

Comment:
```suggestion
    . (zip3 <*> drop 1 <*> drop 2)
```
(Thanks for the cool suggestion @shineli1984 )

Reply:
ah, this is cool. I was looking at it from a `list as applicative functor` perspective, but it looks like it could also be seen as `reader as applicative functor`!

```
zip3 l (drop 1 l) (drop 2 l) === 
(zip3 l) ((drop 1) l) ((drop 2) l) ===
((-> l) zip3) ((-> l) (drop 1) ((-> l) (drop 2)) ===
zip3 <*> drop 1 <*> drop 2
```

Reply:
sorry, mixed this up with the use of `<$>` in ex3. I saw it as reader applicative, but didn't realize that `f <$> id` for reader applicative is simply `f`, because

```
<$> === fmap === .
```